src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java

```diff
@@ -248,75 +250,78 @@ private CompletableFuture<Boolean> checkExistingOrder(CompositePurchaseOrder com
           return completedFuture(compositePurchaseOrder);
         }
         compositePurchaseOrder.getCompositePoLines().get(FIRST_ELEM).setPoLineNumber(poLineNumber);
         return completedFuture(compositePurchaseOrder);
       })
       .exceptionally(t -> {
-        logger.error("Error looking up for existing PO Line Number: {}", compositePurchaseOrder.getPoNumber(), t);
+        LOGGER.error("Error looking up for existing PO Line Number: {}", compositePurchaseOrder.getPoNumber(), t);
         return compositePurchaseOrder;
       });
 
   }
+
   public Void handleError(Throwable throwable) {
-    final javax.ws.rs.core.Response result;
+    Throwable cause = throwable.getCause();
+    LOGGER.error("Exception placing order", cause);
+    asyncResultHandler.handle(Future.succeededFuture(mapExceptionToResponse(cause != null ? cause : throwable)));
+    return null;
+  }
 
-    logger.error("Exception placing order", throwable.getCause());
-    GobiResponse response = new GobiResponse();
-    response.setError(new ResponseError());
+  private static Response mapExceptionToResponse(Throwable t) {
+
+    String message = t.getMessage();
 
-    final Throwable t = throwable.getCause();
     if (t instanceof HttpException) {
-      final int code = ((HttpException) t).getCode();
-      final String message = t.getMessage();
-
-      switch (code) {
-      case 400:
-        response.getError().setCode(CODE_BAD_REQUEST);
-        response.getError().setMessage(HelperUtils.truncate(t.getMessage(), 500));
-        result = respond400WithApplicationXml(GobiResponseWriter.getWriter().write(response));
-        break;
-      case 500:
-        result = respond500WithTextPlain(message);
-        break;
-      case 401:
-        result = respond401WithTextPlain(message);
-        break;
-      default:
-        result = respond500WithTextPlain(message);
+      switch (((HttpException) t).getCode()) {
+        case 400:
+          return respond400WithApplicationXml(writeGobiResponse(CODE_BAD_REQUEST, message));
+        case 401:
+          return respond401WithTextPlain(message);
       }
-    } else if (t instanceof GobiPurchaseOrderParserException) {
-      response.getError().setCode(CODE_INVALID_XML);
-      response.getError().setMessage(HelperUtils.truncate(t.getMessage(), 500));
-      result = respond400WithApplicationXml(GobiResponseWriter.getWriter().write(response));
-    } else {
-      result = respond500WithTextPlain(throwable.getMessage());
     }
 
-    asyncResultHandler.handle(Future.succeededFuture(result));
-    return null;
+    if (t instanceof GobiPurchaseOrderParserException) {
+      return respond400WithApplicationXml(writeGobiResponse(CODE_INVALID_XML, message));
+    }
+
+    if (t instanceof GobiException) {
+      String errorCode = ((GobiException) t).getErrorCode().getCode();
+      return respond500WithApplicationXml(writeGobiResponse(errorCode, message));
+    }
```

Comment:
You can rewrite with java 17
```
if (t instanceof GobiException gobiException) {
    String errorCode = gobiException.getErrorCode().getCode();
    return respond500WithApplicationXml(writeGobiResponse(errorCode, message));
}
```

Reply:
As I checked patterns in switch are supported from v19, can you double check?

Reply:
updating java 17 pr for mod-orders https://github.com/folio-org/mod-orders/pull/760/files for reference

Reply:
We're already using java 17. The expression:
```
return switch (someInt) {
  case 1 -> 1;
  default -> 2;
};
```
is good however following is not:
```
return switch (someObject) {
  case Integer i -> 1;
  case Double d -> 2;
  default -> 3;
};
```